### PR TITLE
Include all dlerror messages to DllNotFoundException

### PR DESCRIPTION
**Pull request by HJLeee**
#70963

### CODE REVIEW

File: src/coreclr/vm/nativelibrary.cpp [hunk 1]
```diff
@@ -107,7 +107,7 @@ namespace
 #if defined(__APPLE__)
             COMPlusThrow(kDllNotFoundException, IDS_EE_NDIRECT_LOADLIB_MAC, libraryNameOrPath.GetUnicode(), GetMessage());
 #elif defined(TARGET_UNIX)
-            COMPlusThrow(kDllNotFoundException, IDS_EE_NDIRECT_LOADLIB_LINUX, libraryNameOrPath.GetUnicode(), GetMessage());
+            COMPlusThrow(kDllNotFoundException, IDS_EE_NDIRECT_LOADLIB_LINUX, libraryNameOrPath.GetUnicode());
```
Comment (jkotas) [inline src/coreclr/vm/nativelibrary.cpp:110]: `IDS_EE_NDIRECT_LOADLIB_LINUX` error message expects two arguments, and this is going to give it only one argument now. I am not sure what the formatting is going to look like, but it is unlikely to be something good.

**Comment by jkotas** [comment]
I expect that there are many situations where the `dlerror` error message is actually helpful. Those cases will be less diagnosable after this change.

If there are situations where the error message returned by `dlerror` is misleading, would it be more appropriate to fix that in `libdl`?

**Comment by HJLeee** [comment]
Narrow down the impact by eliminating overridden error string.

The root cause of incorrect error message is `LoadNativeLibraryBySearch` tries its best for loading p/invoke library.
For a given p/invoke library name `libpinvoke.so`, it attempts 4 times with below name combinations to load the library.
```
currLibNameVariation = {<SBuffer> = {m_size = 34, m_allocation = 82, m_flags = 12, {m_buffer = 0x5583e6df10 "l", m_asStr = 0x5583e6df10 u"libpinvoke.so.so"}},
  static MINIMUM_GUESS = 20, static s_EmptyBuffer = "\000", static s_ACP = 65001, static s_Empty = 0x7fad806290 <s_EmptySpace>}

currLibNameVariation = {<SBuffer> = {m_size = 40, m_allocation = 82, m_flags = 12, {m_buffer = 0x5583e6df10 "l", m_asStr = 0x5583e6df10 u"liblibpinvoke.so.so"}},
  static MINIMUM_GUESS = 20, static s_EmptyBuffer = "\000", static s_ACP = 65001, static s_Empty = 0x7fad806290 <s_EmptySpace>}

currLibNameVariation = {<SBuffer> = {m_size = 28, m_allocation = 82, m_flags = 12, {m_buffer = 0x5583e6df10 "l", m_asStr = 0x5583e6df10 u"libpinvoke.so"}}, static MINIMUM_GUESS = 20,
  static s_EmptyBuffer = "\000", static s_ACP = 65001, static s_Empty = 0x7fad806290 <s_EmptySpace>}

currLibNameVariation = {<SBuffer> = {m_size = 34, m_allocation = 82, m_flags = 12, {m_buffer = 0x5583e6df10 "l", m_asStr = 0x5583e6df10 u"liblibpinvoke.so"}},
  static MINIMUM_GUESS = 20, static s_EmptyBuffer = "\000", static s_ACP = 65001, static s_Empty = 0x7fad806290 <s_EmptySpace>}
```

In this scenario, the last error message for `dlopen()` attempts is kept and it is not possible to determine what error message is beneficial to the developer.
With update, I think error message from dlerror() will be printed when it is not overwritten.

Another possible patch would be keeping error message for exact libname only.

**Comment by jkotas** [comment]
If I am reading the code correctly, this will pretty much always erase the `dlerror` message. There is pretty much always more than one attempt to load library. The first attempt will set the error message, the second attempt will erase it, and so the error message will be lost.

**Comment by jkotas** [comment]
Since we cannot tell which error message is the best one, should keep all error messages (eliminate the duplicates) and include all of them in the exception?

**Comment by HJLeee** [comment]
```
System.DllNotFoundException: Unable to load shared library 'libpinvoke.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libpinvoke.so.so: cannot open shared object file: No such file or directory liblibpinvoke.so.so: cannot open shared object file: No such file or directory /opt/share/libpinvoke.so: wrong ELF class: ELFCLASS32 libpinvoke.so: cannot open shared object file: No such file or directory liblibpinvoke.so: cannot open shared object file: No such file or directory
```

File: src/coreclr/vm/nativelibrary.cpp [hunk 2]
```diff
@@ -137,7 +138,19 @@ namespace
...
         {
+#ifdef TARGET_UNIX
+            //Append dlerror() messages of all attempts
+            SString new_message = SString(SString::Utf8, message);
+            SString::Iterator i = m_message.Begin();
+
+            if (!m_message.Find(i, new_message))
+            {
+                m_message += new_message;
+                m_message += SString(SString::Utf8, " ");
```
Comment (jkotas) [inline src/coreclr/vm/nativelibrary.cpp:149]: I would put each message on a new line.

File: src/coreclr/vm/nativelibrary.cpp [hunk 3]
```diff
@@ -53,6 +53,7 @@ namespace
             LIMITED_METHOD_CONTRACT;
             m_hr = E_FAIL;
             m_priorityOfLastError = 0;
+            m_message = SString(SString::Utf8, "");
```
Comment (jkotas) [inline src/coreclr/vm/nativelibrary.cpp:56]: You can simplify this to `m_message = SString::Empty()` or delete it completely and just depend on the default constructor. The default constructor will initialize the string to empty string.

**Comment by jkotas** [comment]
Could you please make the same change in the managed version at https://github.com/dotnet/runtime/blob/e297470559fb95344ab52c8dad561885a1430e7e/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.NativeAot.Unix.cs#L54 ?

File: src/coreclr/vm/nativelibrary.cpp [hunk 4]
```diff
@@ -137,7 +138,19 @@ namespace
 
         void SetMessage(LPCSTR message)
         {
+#ifdef TARGET_UNIX
+            //Append dlerror() messages of all attempts
+            SString new_message = SString(SString::Utf8, message);
+            SString::Iterator i = m_message.Begin();
+
+            if (!m_message.Find(i, new_message))
```
[the remark this answers is not in the document]
Comment (HJLeee) [reply src/coreclr/vm/nativelibrary.cpp:146]:
```
System.DllNotFoundException: Unable to load shared library 'libpinvoke.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable:    libpinvoke.so.so: cannot open shared object file: No such file or directory    liblibpinvoke.so.so: cannot open shared object file: No such file or directory /opt/share/libpinvoke.so: wrong ELF class: ELFCLASS32  /opt/share/libpinvoke.so: wrong ELF class: ELFCLASS32 libpinvoke.so: cannot open shared object file: No such file or directory    liblibpinvoke.so: cannot open shared object file: No such file or directory
```

If it is not filtered, my example shows `/opt/share/libpinvoke.so: wrong ELF class: ELFCLASS32` twice. (Tested on .NET 6 Linux/aarch64 env)

File: src/coreclr/vm/nativelibrary.cpp (hunk 3 above)
Comment (HJLeee) [reply src/coreclr/vm/nativelibrary.cpp:56]: Thank you for note. Let me fix it.

File: src/coreclr/vm/nativelibrary.cpp (hunk 2 above)
Comment (HJLeee) [reply src/coreclr/vm/nativelibrary.cpp:149]: Thank you for note. Let me fix it.

**Comment by HJLeee** [comment]
I'd love to make a patch for NativeAot.
Currently, I'm having a trouble with setting up NativeAot test env like below.
```
dlerror.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Runtime.linux-x64 with version (= 7.0.0-preview.5.22301.12)
dlerror.csproj : error NU1102:   - Found 1 version(s) in local [ Nearest version: 7.0.0-dev ]
dlerror.csproj : error NU1101: Unable to find package Microsoft.AspNetCore.App.Runtime.linux-x64. No packages exist with this id in source(s): local
```

This is my nuget.config.
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <!--To inherit the global NuGet package sources remove the <clear/> line below -->
    <clear />
    <add key="local" value="${runtime}/artifacts/packages/Release/Shipping" />
    <!--add key="nuget" value="https://api.nuget.org/v3/index.json" /-->
  </packageSources>
</configuration>
```

Please help me setting up the test env.

File: src/coreclr/vm/nativelibrary.cpp [hunk 5]
```diff
@@ -137,7 +138,19 @@ namespace
 
         void SetMessage(LPCSTR message)
         {
+#ifdef TARGET_UNIX
+            //Append dlerror() messages
+            SString new_message = SString(SString::Utf8, message);
+            SString::Iterator i = m_message.Begin();
+            if (!m_message.Find(i, new_message))
+            {
+                m_message += SString(SString::Utf8, " ");
```
Comment (jkotas) [inline src/coreclr/vm/nativelibrary.cpp:147]: Is it intentional to have a space at the start of each line?

Comment (HJLeee) [reply src/coreclr/vm/nativelibrary.cpp:147]: Yes. I thought it's better looking. If you mind, I can remove it.

**Review by jkotas — APPROVED** [review]
LGTM.

@dotnet/interop-contrib Do you have any feedback?

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
> @dotnet/interop-contrib Do you have any feedback?

Nope. Seems reasonable to me. @HJLeee Thank you.

**Comment by HJLeee** [comment]
It's my pleasure and happy to participate in dotnet runtime.

**Pull request merged**